src/fsharp/TypeChecker.fs

```diff
@@ -9431,11 +9430,11 @@ and TcItemThen cenv overallTy env tpenv (item, mItem, rest, afterResolution) del
                         | Some i -> 
                             if isNull(box fittedArgs.[i]) then
                                 fittedArgs.[i] <- arg
-                                let argContainerOpt = match item with
-                                                      | Item.UnionCase(uci, _) -> Some(ArgumentContainer.UnionCase uci)
-                                                      | Item.ExnCase tref -> Some(ArgumentContainer.Type tref)
-                                                      | _ -> None
-                                let argItem = Item.ArgName (argNames.[i], argTys.[i], argContainerOpt)   
+                                let argItem =
+                                    match item with
+                                    | Item.UnionCase (uci, _) -> Item.UnionCaseField (uci, i)
+                                    | Item.ExnCase tref -> Item.RecdField (RecdFieldInfo ([], RFRef (tref, id.idText)))
+                                    | _ -> failwithf "Expecting union case or exception item, got: %O" item   
```

Comment:
What's the danger in this exn (and the others) getting raised? These look to me like the "paint the editor red" kind that we're really keen on not including. The previous code looks like it just hid any problem.

Reply:
It's really not expected to be something else in these particular places, though, I don't know a better way to assert it.
The item is known to be a union case, an exception or an active pattern result from an outer match. The active pattern case is not expected since it cannot have named fields, and we're in the branch processing a resolved named field.

Reply:
I guess we'll find out pretty quickly if something triggers this unexpectedly :)